linux-pipewire: Add PipeWire audio captures

dimtpap opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->
Adds three audio capture sources that utilize PipeWire to capture audio inputs, outputs, as well as audio of specific applications.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open GitHub Issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->
PipeWire is increasingly getting adopted by users as well as major distros (like Fedora), and is on track to be the de facto standard for audio on Linux and others. The PipeWire project has created a PulseAudio-PipeWire bridge, so the current PulseAudio sources work as intended. However, pipewire-pulse has had its issues in the past, although they were minor. Furthermore, PipeWire allows us to capture specific application audio, something that currently needs external tools and tinkering to get working.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->
This has been tested on Arch Linux, on KDE Plasma running X.Org. I am not able to test Wayland captures and PW Audio captures working together since I'm on an Nvidia GPU, and the Nvidia driver is missing features for Wayland to work on it.

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
- New feature (non-breaking change which adds functionality)
<!--- - Tweak (non-breaking change to improve existing functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

Arthus commented:
I'll try to test it this evening using Wayland and Pipewire together. :)

tytan652 commented:
Note: Running OBS with `OBS_USE_EGL=1` allows PipeWire capture being loaded under X11.

dimtpap commented:
> what occurs on audio underruns or overruns with this code, i dont see any callbacks relating (but maybe i missed it).

(This is my first time working with audio, please excuse me if I don't understand something quickly)

No, I haven't done anything related to this problem.
Sometimes when X and Nvidia were being stupid and my system lagged a bit the audio from the sources was crackling (but the sound from my speakers was fine), was that an underrun?

Also I don't know if it matters but it only provides OBS with data when it has data, it doesn't poll the stream for data.

How do other plugins mitigate this problem?

kkartaltepe commented:
> How do other plugins mitigate this problem?

I believe alsa and sndio just drop out in the case of underruns. The pulseaudio plugin attempts to increase latency and grow the audio buffer, but i would caution against reading its code for correctness as its probably not correctly implemented.

> Sometimes when X and Nvidia were being stupid and my system lagged a bit the audio from the sources was crackling (but the sound from my speakers was fine), was that an underrun?

Yes that is the typical result of underruns.

kkartaltepe commented:
I should say that windows also will increase latency to grow the audio buffer (and is implemented correctly). This is our preferred approach as audio dropping (static) in recordings is general less desirable than XXXms of latency.

dimtpap commented:
> I should say that windows also will increase latency to grow the audio buffer (and is implemented correctly). This is our preferred approach as audio dropping (static) in recordings is general less desirable than XXXms of latency.

So to deal with underruns it should request a higher latency from PipeWire then, is that correct?

columbarius commented:
Played a little bit around with this plugin and noticed that the application capture will be silent, if the application is not longer connected to a speaker. I tested vlc and mpd (pulse and pipewire backend) and disconnected them from my speakers with carla. At this point obs didn't record sound anymore. Maybe we are missing a PipeWire property?

dimtpap commented:
> Played a little bit around with this plugin and noticed that the application capture will be silent, if the application is not longer connected to a speaker. I tested vlc and mpd (pulse and pipewire backend) and disconnected them from my speakers with carla. At this point obs didn't record sound anymore. Maybe we are missing a PipeWire property?
> 
> My current guess is that there is no driver in the loop. mpd has the state playing but it won't progress.

Just tried that with Helvum, the stream gets set to paused

dimtpap commented:
> Maybe we are missing a PipeWire property?
Yep, found it (found one?). PW_KEY_NODE_ALWAYS_PROCESS. Maybe there's a more suitable one...?

columbarius commented:
> PW_KEY_NODE_ALWAYS_PROCESS. Maybe there's a more suitable one...?

Nice, but now the stream won't stop, when we switched scenes (I added show and hide callbacks analog to the screencast). Maybe we can ask at #pipewire on IRC.

dimtpap commented:
> > PW_KEY_NODE_ALWAYS_PROCESS. Maybe there's a more suitable one...?
> 
> Nice, but now the stream won't stop, when we switched scenes (I added show and hide callbacks analog to the screencast). Maybe we can ask at #pipewire on IRC.
> 
> But I don't know if the pause on hide is wanted when streaming or not.

Did it even pause before?

I was curious so I looked through the PipeWire source code, seems this flag just sets the wants_driver prop. Maybe try the flag PW_KEY_NODE_WANT_DRIVER instead

dimtpap commented:
Completely reworked the app capture, it now uses a virtual sink to mix all the app streams. It had to be decoupled from the device capture. Tried to keep it DRY as much as possible, but more improvements can be made. Also the device capture logic for connecting to default devices has changed, there wasn't any actual logic to begin with, it was just letting PipeWire decide. I've learned not to say when a piece of software will be finished but I want to believe this is nearly finished.

kkartaltepe commented:
https://github.com/obsproject/obs-studio/pull/6645 did some refactoring to split the portal code from the pipewire code for video streams in case this causes any conflicts as it renames some code. I'm not sure if there is anything that can be shared between the video stream code and this either.

dimtpap commented:
> #6645 did some refactoring to split the portal code from the pipewire code for video streams in case this causes any conflicts as it renames some code. I'm not sure if there is anything that can be shared between the video stream code and this either.

The latest changes were rebased on master just before being pushed and that commit is included. Haven't noticed any conflicts or problems yet. I didn't touch any existing PipeWire code because from what it seems it's just for video as you said. I added everything audio related to the pipewire-audio files but I can see how we can put audio stuff in pipewire.c and .h.

orowith2os commented:
How would this affect having both PulseAudio and PipeWire as audio capture options? Would it be hidden by default, and would appear when an option is enabled in the settings? Or does it JUST add the PipeWire audio capture functionality? 
I believe that this merge should also do the first out of what I mentioned, hiding the PulseAudio capture functionality, so that the menu isn't as cluttered (and as such, not as confusing to read).

dimtpap commented:
> Would it be hidden by default, and would appear when an option is enabled in the settings? Or does it JUST add the PipeWire audio capture functionality? I believe that this merge should also do the first out of what I mentioned, hiding the PulseAudio capture functionality

This PR only adds the PipeWire audio capture sources. I feel like PulseAudio vs PipeWire source loading should be done in a separate PR, once this is merged

orowith2os commented:
This PR hasn't had any activity in a while, and the plugin version seems to work fine for me, from my testing. Will this be merged anytime soon?

kkartaltepe commented:
> This PR hasn't had any activity in a while, and the plugin version seems to work fine for me, from my testing. Will this be merged anytime soon?

I assume this is because the author had no reason to update it. As its still in draft i wouldn't expect it to be reviewed for merge until the author marks it as ready. We are also in the middle of releasing v28 so its very unlikely anyone will review for a while even if it were undrafted.

dimtpap commented:
I'm waiting for https://github.com/obsproject/obs-studio/pull/6669

columbarius commented:
@dimtpap Might you take a look at #6669 and tell me if you would just reuse the core/obs_pw api, or if you would include the audio stuff into the stream api?

dimtpap commented:
I was told by Columbarius that work on the unified wrappers has stalled, and since this is ready for general use I can go ahead mark this as ready

Phr3d13 commented:
Would love to see this included in OBS. adding this manually is a pain on a Steam Deck.

cboin1996 commented:
Is this still planning on being added?

Fenrirthviti commented:
>  Is this still planning on being added?

This is still planned, but there is a lot of moving parts. You can follow the discussion and next steps here: https://github.com/obsproject/obs-studio/discussions/7998

cboin1996 commented:
> > Is this still planning on being added?
> 
> This is still planned, but there is a lot of moving parts. You can follow the discussion and next steps here: #7998

Thanks for the reply! Will do

Plarpoon commented:
Hi, how is this PR going? I see that not much is moving and the linked issues are either all resolved or waiting for approval. This is a great addition to OBS and I think it should be merged as soon as ready. Taking the opportunity to thank everyone involved in the project too!

dimtpap commented:
> Hi, how is this PR going? I see that not much is moving and the linked issues are either all resolved or waiting for approval.
Feature wise it's complete.
There are ongoing efforts to refactor OBS's PipeWire code to keep it cleaner but are not done. This PR will be refactored to adapt to those changes.

Plarpoon commented:
That's absolutely great news to hear @dimtpap, this one is a must for proper Linux support and considering Wayland adoption is on the rise it's fantastic to know that OBS is prepared for it!

v4u6h4n commented:
Thanks to everyone who worked on this, really looking forward to this :-)

dimtpap commented:
Marked as draft as there's a need for a sandbox-friendly way to access system audio. There's no portal for that yet, so until that happens this needs the `--filesystem=xdg-run/pipewire-0` hole, which is not wanted for OBS.

Bleuzen commented:
> this needs the `--filesystem=xdg-run/pipewire-0` hole, which is not wanted for OBS.

OBS already has this:
https://github.com/obsproject/obs-studio/blob/409bd12db38095f440bd59da68a7df46c6787a4e/build-aux/com.obsproject.Studio.json#L14

So unless it is planned to be removed in the next OBS version this doesn't look like a blocker

dimtpap commented:
> > this needs the `--filesystem=xdg-run/pipewire-0` hole, which is not wanted for OBS.
> 
> OBS already has this:
> 
> So unless it is planned to be removed in the next OBS version this doesn't look like a blocker

It's there because of some PulseAudio/JACK audio stuff I think. It's unclear if it will be removed at any future version. With this PR it will be required to work, but the focus is on PipeWire+portals and dynamic permissions and it should be removed. I marked it as draft after asking in the dev channel of the Discord server and got the ok from a maintainer. It was marked as ready for merging for a while without any activity probably because of those reasons and drafting should better reflect the need for other components to exist. Of course if the OBS team wants this to move forward I can undraft and rebase etc.

GloriousEggroll commented:
seems this may be busted with recent betas (30.1.2):

```
error: os_dlopen(/usr/lib64/obs-plugins/linux-pipewire.so->/usr/lib64/obs-plugins/linux-pipewire.so): /usr/lib64/obs-plugins/linux-pipewire.so: undefined symbol: pipewire_audio_capture_app_load

error: os_dlopen(/usr/lib64/obs-plugins/linux-pipewire.so->/usr/lib64/obs-plugins/linux-pipewire.so): /usr/lib64/obs-plugins/linux-pipewire.so: undefined symbol: pipewire_audio_capture_app_load

```

kkartaltepe commented:
> seems this may be busted with recent betas (30.1.2):
> 
> ```
> error: os_dlopen(/usr/lib64/obs-plugins/linux-pipewire.so->/usr/lib64/obs-plugins/linux-pipewire.so): /usr/lib64/obs-plugins/linux-pipewire.so: undefined symbol: pipewire_audio_capture_app_load
> 
> error: os_dlopen(/usr/lib64/obs-plugins/linux-pipewire.so->/usr/lib64/obs-plugins/linux-pipewire.so): /usr/lib64/obs-plugins/linux-pipewire.so: undefined symbol: pipewire_audio_capture_app_load
> ```

PRs with the old build system are probably not going to work, you would need to update it to the new cmake tooling since the old stuff is also being deleted eventually.

XenHat commented:
Can I look forward to getting this added this year? This would simplify many people's workflow, including mine.

mihawk90 commented:
> Can I look forward to getting this added this year?

Doubtful since this is not entirely up to OBS. As noted above this depends largely on the existence and implementation of an audio portal, then for that implementation to trickle down into each of the Desktop Environments, then for those DE-implementations to trickle down to individual distros. In addition to that OBS itself also needs some work for it done.

In the meantime, @dimtpap has this available as a plugin:
https://github.com/dimtpap/obs-pipewire-audio-capture

h8d13 commented:
Bump <3

h8d13 commented:
Doesn't gnome also default to wayland xdg and pipewire, just like KDE ? I'm not sure I'm following you.

Plarpoon commented:
I think you are making a bit of confusion, I strongly suggest you go read up what each of the words I am about to write do from the actual official pages but I am going to give a quick explanation so other people that will come here in the future will not make the same mistakes.

- Wayland is a display protocol, it keeps windows isolated by nature, ignore this.

- KDE/GNOME or whatever else are compositors that implement the wayland protocol, they decide how to decorate a window or how to make it behave.

- XDG is a group of communities/project/developers that set together standards so that apps behave roughly the same way across all of the different setups.

- Portals are an interface between applications that implement the standards chosen by the XDG group. For example if your app has an authorization to record your microphone or send audio to the multimedia engine and so on.

- PipeWire is just the actual engine that carries the bistream of audio/video from the compositor to the apps.



If you manually install this plugin it will work, the issue is not a technical one but a political one. OBS cannot ship it by default because that would mean deciding that the way the audio is accessed is allowed, but that is something that the XDG group has yet to define as a standard. Until such a standard is defined OBS will likely not take on itself the responsibility of shipping by default something non standard.

It's a fantastic thing what the XDG group is doing, but it is notoriously slow on some aspects, so you just have to wait that out or push the various communities to find an agreement.

Fenrirthviti commented:
I'd like to ask a bit of an uncomfortable question:

Is there value to the overall initiative in keeping this PR open? It is effectively stale, and hasn't had updates in around 2 years. It acts as a lightning rod for "why isn't this merged yet?" discussions, which generally just go in circles of Linux politics and other upstream details that are not ready yet.

I can say that we are certainly interested in this as a feature, but until the ecosystem is ready for it, I don't see much value in leaving it open on our repo and inviting comments asking why it's not merged yet.

A reminder to those watching this issue that dimtpap has released a [plugin version of this PR](https://github.com/dimtpap/obs-pipewire-audio-capture), which can be used in the meantime and is being kept up-to-date

luxaritas commented:
Would it be helpful to edit the description with a "merging this is blocked by x" perhaps with a link to other active discussion to follow (if any)?
